\documentclass[12pt]{article}
%\usepackage[english]{babel}
%\usepackage[utf8x]{inputenc}
\usepackage{fullpage}
\usepackage{float}

\usepackage{listings}
\usepackage{color}

\title{Supplemental Information}
\author{Pedersen \textit{et al.}}

\date{}

\definecolor{mygreen}{rgb}{0,0.6,0}
\definecolor{mygray}{rgb}{0.5,0.5,0.5}
\definecolor{mymauve}{rgb}{0.58,0,0.82}

\lstset{ %
  backgroundcolor=\color{white},   % choose the background color
  basicstyle=\footnotesize,        % size of fonts used for the code
  breaklines=true,                 % automatic line breaking only at whitespace
  captionpos=b,                    % sets the caption-position to bottom
  commentstyle=\color{mygreen},    % comment style
  escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  keywordstyle=\color{blue},       % keyword style
  stringstyle=\color{mymauve},     % string literal style
}
\usepackage{graphicx}

\begin{document}

\maketitle

\section{Aligners Compared}

\begin{table}[!t]
\processtable{Alignment Methods Compared\label{Tab:01}}
{\begin{tabular}{lll}\toprule
software & version & command\\\midrule
bismark & 0.10.1 & bismark --gzip --maxins 1000 -n 3 -l 20 --bam\\
bsmap & 2.74 & bsmap -s12 -v3 -m0 -x1000 -S42 -n0 -s12 -I1\\
bwa-meth & 0.06 & bwa-meth\\
gsnap & 2014-02-28 & gsnap --npaths 1 --quiet-if-excessive\\
last & 392 & last-bisulfite-paired.sh\\\botrule
\end{tabular}}{We used a modified version of the calling script for last}
\end{table}


\section{Comparison On Real Data}

We compared 5 aligners on real data as described in the text. Supplemental
Figure 1 below shows a color version of Figure 1 from the main text.

\begin{figure}[H]%supplemental figure1
    \centerline{\includegraphics[width=86mm]{supp-1-real-qual-summary.eps}}
    \caption{Percent of paired-end, 100-base reads on (y) and off (x) target for the tested aligners. Aligners that report mapping quality are shown as connected dots for each quality cut-off. Reads are limited to those considered as primary, mapped alignments by the aligner. This is a color version of figure 1
from the paper}\label{suppfig:01}
\end{figure}

When we trim the reads with \emph{Sickle} before aligning, the result is shown below in Supplemental Figure 2. We report the percent of reads aligned relative
to the original count in the untrimmed data because we are interested in the
overall mapping rate.

%\begin{figure}[!htpb]%figure1
\begin{figure}[H]%figure1
    \centerline{\includegraphics[width=86mm]{supp-1-real-trim-qual-summary.eps}}
    \caption{Percent of paired-end, \emph{trimmed} 100-base reads on (y) and off (x) target for the tested aligners. Aligners that report mapping quality are shown as connected dots for each quality cut-off. Reads are limited to those considered as primary, mapped alignments by the aligner.
from the paper}\label{suppfig:02}
\end{figure}

\subsection{Resources}

%http://truben.no/latex/table/
\begin{table}[H]
    \centering
    \caption{Resources on real data}
    \begin{tabular}{lllll} \hline
    trimmed & program & time(min) & mem(GB) & dataset \\ \hline
    no      & bis1    & 573.18    & 8.30    & real    \\
    no      & bsmap   & 7994.27   & 23.11   & real    \\
    no      & bwa     & 356.04    & 15.44   & real    \\
    no      & gsnap   & 3393.78   & 10.98   & real    \\
    no      & last    & 621.75    & 34.53   & real    \\
    yes     & bis1    & 584.45    & 8.28    & real    \\
    yes     & bsmap   & 4941.04   & 23.29   & real    \\
    yes     & bwa     & 341.75    & 16.75   & real    \\
    yes     & gsnap   & 2779.10   & 12.12   & real    \\
    yes     & last    & 536.70    & 34.53   & real    \\
    \end{tabular}
\end{table}

\section{Comparison On Simulated Data}
Paired-end, 100-base reads were simulated using the tool \emph{Sherman} (http://www.bioinformatics.babraham.ac.uk/projects/sherman/) and aligned using the
same parameters as for Figure 1 for the main paper.
Supplemental Figure 3 below shows the result for these simulations.


\begin{figure}[H]%figure3
    \centerline{\includegraphics[width=86mm]{supp-2-sim-qual-summary.eps}}
    \caption{Percent of paired-end, \emph{simulated} 100-base reads on (y) and off (x) target for the tested aligners. Aligners that report mapping quality are shown as connected dots for each quality cut-off. Reads are limited to those considered as primary, mapped alignments by the aligner.
from the paper}\label{suppfig:03}
\end{figure}

Supplemental Figure 4 below shows the result for trimmed and simulated reads.
Trimming removes some read-pairs if one or both reads had low-quality. We
report the percent of reads aligned relative to the original, un-trimmed number
since we are interested in the overall mapping rate.

\begin{figure}[H]%figure3
    \centerline{\includegraphics[width=86mm]{supp-2-sim-trim-qual-summary.eps}}
    \caption{Percent of paired-end, \emph{trimmed}, \emph{simulated} 100-base reads on (y) and off (x) target for the tested aligners. Aligners that report mapping quality are shown as connected dots for each quality cut-off. Reads are limited to those considered as primary, mapped alignments by the aligner.
from the paper}\label{suppfig:04}
\end{figure}

\subsection{Resources}

%http://truben.no/latex/table/
\begin{table}[H]
    \centering
    \caption{Resources on simulated data}
    \begin{tabular}{lllll} \hline
    trimmed & program & time(min) & mem(GB) & dataset \\ \hline
    no      & bis1    & 62.00     & 8.27    & sim     \\
    no      & bsmap   & 131.92    & 22.93   & sim     \\
    no      & bwa     & 122.05    & 16.47   & sim     \\
    no      & gsnap   & 244.82    & 9.50    & sim     \\
    no      & last    & 131.17    & 25.90   & sim     \\
    yes     & bis1    & 62.13     & 8.27    & sim     \\
    yes     & bsmap   & 65.00     & 22.92   & sim     \\
    yes     & bwa     & 160.39    & 20.98   & sim     \\
    yes     & gsnap   & 33.21     & 9.55    & sim     \\
    yes     & last    & 144.46    & 25.90   & sim     \\
    \end{tabular}
\end{table}


\section{Mapping and Trimming}
We note that while trimming seems to improve accuracy for simulated data,
it does not improve anything for real data. This could be because
the reads from the targetted protocol are likely to come from CpG-rich regions
with simple sequence and trimming them increases the possiblity for mis-maps
and reduces the likelihood of a unique mapping.

\section{\textit{bwa-meth} Installation And Requirements}

\textit{Bwa-meth} depends on samtools and a single python library, \textit{toolshed}.
The latter can be installed by running \emph{python setup.py install} from the main
directory of the \textit{bwa-meth} project. Samtools is a C library installed on most
systems and available at https://github.com/samtools/samtools.

For tabulation of methylation by CpG, \emph{Bis-SNP} \cite{bissnp} is required.
The java .jar file is available from: http://sourceforge.net/projects/bissnp/files/

For CNV detection from BS-Seq data, the R package \emph{cn.mops} \cite{cnmops} is required.
It is available
from bioconductor \cite{bioconductor} at: http://bioconductor.org/packages/devel/bioc/html/cn.mops.html

\section{Advantages}
\textit{bwa-meth} provides a number of advantages:

1. ease of installation -- requires only python (python2 or python3), bwa,
   and samtools for the aligner.

2. easy of use -- a single script indexes the fasta, aligns the reads, and
   tabulates methylation. Default parameters work well.

3. speed -- due to the efficient parallelization by bwa, bwa-meth can run on
   as many CPU's are available on a single machine.

4. disk-usage -- bwa-meth can take either compressed or uncompressed files
   and the reads are streamed directly to the aligner, never written to disk. An
   aligner that requires trimmed reads and that writes the converted reads to
   disk will require 3X as much storage for just the raw sequence data.

5. useful output -- the BAM that results from bwa-meth is sorted, indexed and
   contains the proper mapping quality scores, alignment flags, and read-groups.
   In addition, the header @SQ lines are sorted so that tools like Picard and
   GATK will accept them as is.

6. simplicity -- we have carefully designed bwa-meth to be quite simple and
   the code readable so that enhancements are simple. For example, see \#7.

7. strand-specificity -- After we posted bwa-meth on seqanswers, Devon Ryan,
   the author of Bison, noted that many capture methods target only the reverse
   strand -- commonly called the original bottom. Though this would be simple to
   add to any aligner, ours is the only one that currently supports this. See
   from the supplemental figures the reduction in off-target reads that results
   from only considering reads that align to the original bottom strand.

8. specificity and sensitivity -- while the difference between aligners in
   this regard is small, we have shown that bwa-meth consistently peforms as the
   best or among the best. We believe this to be the case in general and not
   limited to the specific cases that we have shown.

9. accuracy without trimming -- the accuracy is due in part to bwa's local
   alignment. This is something that is not supported in bismark and means that
   we can achieve very good accuracy without performing additional trimming
   which would result in additional disk-usage and processing time.

\section{Additional Features}
\textit{Bwa-meth} and GSNAP are the only program that outputs a BAM file that passes picard's ValidateSam without errors.
Last does not report the proper pair information in all cases and none of the other aligners add a read-group.
For the comparison, we added sorting and forced SAM output by the other aligners regardless of their default.
\textit{Bwa-meth} outputs a read-group for each sample by default and allows that to be customized.

\textit{bwa-meth} can calculate bias of methylation estimates by location in the read:

\begin{lstlisting}[language=bash]
python bias-plot.py input.bam ref.fa
\end{lstlisting}
This will create a bias-plot showing bases in the reads that should not be considered
for methylation calls.

\textit{Bwa-meth} defers tabulation of the methylation scores to Bis-SNP \cite{bissnp} by offering a simplified interface:
\begin{lstlisting}[language=bash]
bwa-meth tabulate \
    --trim 3,3 \
    --map-q 30 \
    --bissnp BisSNP-0.82.2.jar \
     --reference /path/to/ref.fasta \
     input.bam
\end{lstlisting}
Where the arguments are sent to Bis-SNP to, for example trim the first and last 3 bases
from each read to avoid bias.

A full example on real data is at:
https://github.com/brentp/bwa-meth/tree/master/example/

%\begin{thebibliography}{}
\bibliographystyle{abbrv}
    \bibliography{document}
%\end{thebibliography}

\end{document}
